src/vulkan-renderer/wrapper/device.cpp

```diff
@@ -207,43 +459,37 @@ Device::Device(const wrapper::Instance &instance, const VkSurfaceKHR surface, bo
         m_transfer_queue_family_index = m_graphics_queue_family_index;
     }
 
-    std::vector<const char *> device_extensions_wishlist = {
-        // Since we want to draw on a window, we need the swapchain extension.
-        VK_KHR_SWAPCHAIN_EXTENSION_NAME,
-    };
+    VkPhysicalDeviceFeatures available_features{};
+    vkGetPhysicalDeviceFeatures(physical_device, &available_features);
 
-#ifndef NDEBUG
-    if (enable_vulkan_debug_markers) {
-        // Debug markers allow the assignment of internal names to Vulkan resources.
-        // These internal names will conveniently be visible in debuggers like RenderDoc.
-        // Debug markers are only available if RenderDoc is enabled.
-        device_extensions_wishlist.push_back(VK_EXT_DEBUG_MARKER_EXTENSION_NAME);
-    }
-#endif
+    const auto comparable_required_features = get_device_features(required_features);
+    const auto comparable_optional_features = get_device_features(optional_features);
+    const auto comparable_available_features = get_device_features(available_features);
 
-    std::vector<const char *> enabled_device_extensions;
+    constexpr auto FEATURE_COUNT = sizeof(VkPhysicalDeviceFeatures) / sizeof(VkBool32);
+    std::vector<VkBool32> features_to_enable(FEATURE_COUNT, VK_FALSE);
 
-    for (const auto &device_extension_name : device_extensions_wishlist) {
-        if (is_extension_supported(m_physical_device, device_extension_name)) {
-            spdlog::trace("Device extension {} is available on this system", device_extension_name);
-            enabled_device_extensions.push_back(device_extension_name);
-        } else {
-            throw std::runtime_error("Device extension " + std::string(device_extension_name) +
-                                     " is not available on this system!");
+    for (std::size_t i = 0; i < FEATURE_COUNT; i++) {
+        if (comparable_required_features[i] == VK_TRUE) {
+            features_to_enable[i] = VK_TRUE;
+        }
+        if (comparable_optional_features[i] == VK_TRUE && comparable_available_features[i] == VK_TRUE) {
+            features_to_enable[i] = VK_TRUE;
+        }
+        if (comparable_optional_features[i] == VK_TRUE && comparable_available_features[i] == VK_FALSE) {
+            spdlog::warn("The physical device {} does not support {}!", get_physical_device_name(physical_device),
+                         get_feature_description(i));
         }
```

Comment:
```suggestion
        if (comparable_optional_features[i] == VK_TRUE) {
            if (comparable_available_features[i] == VK_TRUE) {
                features_to_enable[i] = VK_TRUE;
            } else {
                spdlog::warn("The physical device {} does not support {}!", get_physical_device_name(physical_device),
                             get_feature_description(i));
            }
        }
```

Reply:
I also completely forgot about the documentation of `required_extensions` and `physical_device`. Done.

Reply:
Done.